File: frontend/pages/class_detail.py
import streamlit as st

from services.class_api import ClassApiError, delete_class, get_class
from services.student_api import StudentApiError, create_student, list_students


WEEKDAY_LABELS = {
    "mon": "Monday",
    "monday": "Monday",
    "tue": "Tuesday",
    "tues": "Tuesday",
    "tuesday": "Tuesday",
    "wed": "Wednesday",
    "weds": "Wednesday",
    "wednesday": "Wednesday",
    "thu": "Thursday",
    "thur": "Thursday",
    "thurs": "Thursday",
    "thursday": "Thursday",
    "fri": "Friday",
    "friday": "Friday",
    "sat": "Saturday",
    "saturday": "Saturday",
    "sun": "Sunday",
    "sunday": "Sunday",
}


def _safe_text(value: object, fallback: str = "N/A") -> str:
    if value is None:
        return fallback
    text = str(value).strip()
    return text if text else fallback


def _display_day_label(day: object) -> str:
    normalized = _safe_text(day, fallback="").strip()
    if not normalized:
        return ""
    return WEEKDAY_LABELS.get(normalized.lower(), normalized)


def _format_mood_label(value: object) -> str:
    mood = _safe_text(value, fallback="No mood yet")
    if mood.lower() in {"n/a", "no mood yet"}:
        return mood
    return mood[:1].upper() + mood[1:].lower()


def _render_metric_tile(label: str, value: str) -> None:
    st.markdown(
        f"<div class='analysis-metric'><div class='analysis-metric-label'>{label}</div><div class='analysis-metric-value'>{value}</div></div>",
        unsafe_allow_html=True,
    )


def _render_header(classroom: dict) -> None:
    schedule_days = classroom.get("schedule_days") or []
    schedule_chips = []
    for day in schedule_days:
        label = _display_day_label(day)
        if label:
            schedule_chips.append(f"<span class='analysis-chip'>{label}</span>")

    if not schedule_chips:
        schedule_chips = ["<span class='analysis-chip'>No schedule set</span>"]

    subtitle = f"{_safe_text(classroom.get('grade_age_group'))} • {_safe_text(classroom.get('description'), fallback='No class description added yet.')}"

    st.markdown(
        f"""
        <div class='analysis-hero'>
            <h2 style='margin:0'>{_safe_text(classroom.get('class_name'))}</h2>
            <p class='analysis-subtitle'>{subtitle}</p>
            <div class='analysis-chip-row'>
                {''.join(schedule_chips)}
            </div>
        </div>
        """,
        unsafe_allow_html=True,
    )


def _student_card(student: dict) -> bool:
    student_id = student.get("id")
    mood = _format_mood_label(student.get("last_predicted_mood"))
    last_predicted = _safe_text(student.get("last_predicted_label"), fallback="No predictions yet")
    total_analyses = _safe_text(student.get("total_analyses"), fallback="0")

    st.markdown(
        f"""
        <div class='student-click-wrap'>
            <div class='student-grid-card'>
                <div class='student-grid-head'>
                    <div class='student-avatar'>👧</div>
                    <div class='student-head-copy'>
                        <div class='student-grid-name'>{_safe_text(student.get('name'))}</div>
                        <div class='student-grid-mood'>{mood}</div>
                    </div>
                </div>
                <div class='student-grid-footer'>
                    <span class='student-grid-time'>{last_predicted}</span>
                    <span class='student-grid-analyses'>{total_analyses} analyses</span>
                </div>
            </div>
        </div>
        """,
        unsafe_allow_html=True,
    )

    if st.button("Open Student Profile", key=f"student_open_{student_id}", use_container_width=True):
        st.session_state.selected_student = student
        st.session_state.page = "student_profile"
        st.rerun()
        return True

    return False

def class_detail_page():
    token = st.session_state.get("auth_token")

    cls = st.session_state.get("selected_class")

    if not cls:
        st.warning("No class selected")
        return

    class_id = cls.get("id")
    if not isinstance(class_id, int):
        st.error("Selected class is missing a valid id.")
        return

    try:
        cls = get_class(class_id, token=token)
        st.session_state.selected_class = cls
    except ClassApiError as exc:
        st.warning(f"Could not refresh class details from backend: {exc}")

    _render_header(cls)

    st.markdown("<h4 class='analysis-section-title'>Class Actions</h4>", unsafe_allow_html=True)
    action_col1, action_col2, action_col3 = st.columns(3)
    with action_col1:
        if st.button("Back to Classes", key="detail_back_to_classes", use_container_width=True):
            st.session_state.page = "My Classes"
            st.rerun()

    with action_col2:
        if st.button("Edit Class", key="detail_edit_class", use_container_width=True):
            st.session_state.page = "edit_class"
            st.rerun()

    with action_col3:
        if st.button("Delete Class", key="detail_delete_class", use_container_width=True):
            st.session_state.delete_confirm_class_id = class_id

    if st.session_state.get("delete_confirm_class_id") == class_id:
        st.warning("Delete this class permanently? This action cannot be undone.")
        confirm_col, cancel_col = st.columns([1, 1])
        with confirm_col:
            if st.button("Yes, Delete", key="detail_delete_confirm", use_container_width=True):
                try:
                    delete_class(class_id, token=token)
                except ClassApiError as exc:
                    st.error(f"Failed to delete class: {exc}")
                else:
                    st.session_state.delete_confirm_class_id = None
                    st.session_state.selected_class = None
                    st.session_state.page = "My Classes"
                    st.success("Class deleted successfully.")
                    st.rerun()
        with cancel_col:
            if st.button("Cancel", key="detail_delete_cancel", use_container_width=True):
                st.session_state.delete_confirm_class_id = None
                st.rerun()

    students: list[dict] = []
    try:
        students = list_students(class_id, token=token)
    except StudentApiError as exc:
        st.warning(f"Could not load students from backend: {exc}")

    st.divider()

    col1, col2, col3 = st.columns(3)
    total_analyses = sum(int(s.get("total_analyses") or 0) for s in students)

    with col1:
        _render_metric_tile("Total Students", str(len(students)))

    with col2:
        _render_metric_tile("Total Analyses", str(total_analyses))

    with col3:
        recent_count = len([s for s in students if _safe_text(s.get("last_predicted_label")) in {"just now"}])
        _render_metric_tile("Updated Recently", str(recent_count))

    st.divider()

    header_col, action_col = st.columns([3.8, 1.2])
    with header_col:
        st.subheader(f"Students ({len(students)})")
    with action_col:
        if st.button("Add New Student", key="detail_add_student_toggle", use_container_width=True):
            st.session_state.show_add_student_form = not st.session_state.get("show_add_student_form", False)

    if st.session_state.get("show_add_student_form", False):
        st.markdown("<div class='card'>", unsafe_allow_html=True)
        with st.form("add_student_form", clear_on_submit=True):
            student_name = st.text_input("Student Name", placeholder="e.g. Emma Watson")
            student_age_group = st.text_input("Age Group", placeholder="e.g. Grade 4")
            submitted = st.form_submit_button("Create Student", type="primary")

        st.markdown("</div>", unsafe_allow_html=True)

        if submitted:
            if not student_name.strip() or not student_age_group.strip():
                st.error("Student name and age group are required.")
            else:
                try:
                    create_student(
                        class_id=class_id,
                        name=student_name.strip(),
                        age_group=student_age_group.strip(),
                        token=token,
                    )
                except StudentApiError as exc:
                    st.error(f"Failed to create student: {exc}")
                else:
                    st.success("Student added successfully.")
                    st.session_state.show_add_student_form = False
                    st.rerun()

    search = st.text_input("Search students", placeholder="Type student name...")

    if search:
        students = [s for s in students if search.lower() in _safe_text(s.get("name"), fallback="").lower()]

    if not students:
        st.caption("No students found for this class yet.")
        return

    columns = st.columns(3, gap="medium")
    for index, student in enumerate(students):
        with columns[index % 3]:
            _student_card(student)

File: frontend/js/pages/class-detail.js
import { createShell, attachShellHandlers } from '../components/shell.js';
import { fetchClassById } from '../api/classes.js';
import { deleteClass } from '../api/classes.js';
import { showToast } from '../components/toast.js';

export async function renderPage({ params } = {}) {
	const appElement = document.getElementById('app');
	if (!appElement) return;

	const classId = params?.id;
	if (!classId) {
		showToast('error', 'Class not found.');
		window.location.hash = '#/classes';
		return;
	}

	try {
		const classroom = await fetchClassById(classId);
		renderClassDetail(appElement, classroom);
	} catch (error) {
		showToast('error', error.message || 'Failed to load class details.');
		renderErrorState(appElement, error);
	}
}

function renderClassDetail(appElement, classroom) {
	const students = Array.isArray(classroom.students) ? classroom.students : [];
	const totalAnalyses = students.reduce((sum, student) => sum + Number(student.total_analyses || 0), 0);
	const recentUpdates = students.filter((student) => Boolean(student.last_predicted_at)).length;

	const pageContent = `
		<div class="dashboard-container class-detail-page">
			<section class="class-detail-hero card">
				<div class="class-detail-hero-top">
					<div class="class-detail-hero-identity">
						<div class="class-detail-back-link-wrap">
							<a href="#/classes" class="btn btn-ghost btn-sm">
								<i data-lucide="arrow-left"></i>
								Back
							</a>
						</div>
						<div class="class-detail-hero-info">
							<h2>${safeText(classroom.class_name || 'Class')}</h2>
							<p class="class-detail-meta">${safeText(classroom.grade_age_group || 'N/A')}${classroom.description ? ` · ${safeText(classroom.description)}` : ''}</p>
							<div class="class-detail-chips">
								${renderScheduleDays(classroom.schedule_days)}
							</div>
						</div>
					</div>
					<div class="class-detail-actions">
						<a href="#/classes/${classroom.id}/edit" class="btn btn-secondary">
							<i data-lucide="pencil"></i>
							Edit Class
						</a>
						<button id="delete-class-btn" class="btn btn-danger-ghost" type="button">
							<i data-lucide="trash-2"></i>
							Delete Class
						</button>
					</div>
				</div>
			</section>

			<section class="classes-summary-grid class-detail-summary-grid">
				<article class="glass-card stat-card">
					<div class="stat-card-icon-row">
						<div class="stat-card-icon"><i data-lucide="users"></i></div>
					</div>
					<div class="stat-card-value">${students.length}</div>
					<div class="stat-card-label">Total Students</div>
				</article>

				<article class="glass-card stat-card">
					<div class="stat-card-icon-row">
						<div class="stat-card-icon"><i data-lucide="scan-eye"></i></div>
					</div>
					<div class="stat-card-value">${totalAnalyses}</div>
					<div class="stat-card-label">Total Analyses</div>
				</article>

				<article class="glass-card stat-card">
					<div class="stat-card-icon-row">
						<div class="stat-card-icon"><i data-lucide="clock-3"></i></div>
					</div>
					<div class="stat-card-value">${recentUpdates}</div>
					<div class="stat-card-label">Recently Updated</div>
				</article>
			</section>

			<section class="card class-detail-students-section">
				<div class="classes-page-header-row">
					<div>
						<h3>Students (${students.length})</h3>
						<p class="text-muted" style="margin-top: 6px;">Browse student profiles and their latest analysis summary.</p>
					</div>
					<a href="#/classes/${classroom.id}/students/new" class="btn btn-primary btn-sm">
						<i data-lucide="plus"></i>
						Add Student
					</a>
				</div>

				<div class="classes-toolbar class-detail-toolbar">
					<div class="search-wrapper classes-search-wrapper">
						<i data-lucide="search"></i>
						<input id="students-search" type="search" class="form-input" placeholder="Search students by name or mood" />
					</div>
				</div>

				<div id="students-grid" class="class-detail-students-grid">
					${renderStudentCards(classroom.id, students)}
				</div>
			</section>
		</div>
	`;

	appElement.innerHTML = createShell(pageContent, { topbarTitle: safeText(classroom.class_name || 'Class Detail') });
	attachShellHandlers();
	bindStudentSearch(classroom.id, students);
	bindDeleteAction(classroom);

	if (window.lucide && typeof window.lucide.createIcons === 'function') {
		window.lucide.createIcons();
	}
}

function bindStudentSearch(classId, students) {
	const searchInput = document.getElementById('students-search');
	const studentsGrid = document.getElementById('students-grid');
	if (!searchInput || !studentsGrid) return;

	searchInput.addEventListener('input', () => {
		const query = searchInput.value.trim().toLowerCase();
		const filteredStudents = students.filter((student) => {
			const name = String(student.name || '').toLowerCase();
			const mood = String(student.last_predicted_mood || '').toLowerCase();
			return name.includes(query) || mood.includes(query);
		});

		studentsGrid.innerHTML = renderStudentCards(classId, filteredStudents);
		if (window.lucide && typeof window.lucide.createIcons === 'function') {
			window.lucide.createIcons();
		}
	});
}

function bindDeleteAction(classroom) {
	const deleteButton = document.getElementById('delete-class-btn');
	if (!deleteButton) return;

	deleteButton.addEventListener('click', async () => {
		openModal(`
			<div class="modal-confirm-body">
				<h3>Delete ${safeText(classroom.class_name)}?</h3>
				<p class="text-muted">This will remove the class and keep its students unavailable in the current teacher workspace.</p>
				<div class="modal-actions" style="margin-top: 20px;">
					<button id="cancel-delete-class-btn" class="btn btn-ghost" type="button">Cancel</button>
					<button id="confirm-delete-class-btn" class="btn btn-danger" type="button">Delete</button>
				</div>
			</div>
		`);
	});
		const cancelButton = document.getElementById('cancel-delete-class-btn');
		const confirmButton = document.getElementById('confirm-delete-class-btn');
		cancelButton?.addEventListener('click', () => closeModal());
		confirmButton?.addEventListener('click', async () => {
			try {
				confirmButton.disabled = true;
				await deleteClass(classroom.id);
				showToast('success', 'Class deleted successfully.');
				closeModal();
				window.location.hash = '#/classes';
			} catch (error) {
				confirmButton.disabled = false;
				showToast('error', error.message || 'Failed to delete class.');
			}
		});
}

function renderStudentCards(classId, students) {
	if (!students.length) {
		return `
			<div class="classes-empty-state card">
				<i data-lucide="user"></i>
				<h3>No students yet</h3>
				<p class="text-muted">Add the first student to this class to start tracking analyses.</p>
			</div>
		`;
	}

	return students
		.map((student) => {
			const initials = getInitials(student.name);
			const moodBadgeClass = getMoodBadgeClass(student.last_predicted_mood);
			return `
				<a href="#/classes/${classId}/students/${student.id}" class="card card-clickable class-student-card">
					<div class="class-student-card-top">
						<div class="student-avatar-circle">${safeText(initials)}</div>
						<div class="class-student-card-info">
							<h4>${safeText(student.name || 'Unnamed student')}</h4>
							<p class="text-muted text-sm">Joined ${formatDateShort(student.joined_at)}</p>
						</div>
					</div>

					<div class="class-student-card-meta">
						${student.last_predicted_mood ? `<span class="badge ${moodBadgeClass}">${safeText(student.last_predicted_mood)}</span>` : '<span class="badge badge-gray">No mood yet</span>'}
						<span class="badge badge-gray">${Number(student.total_analyses || 0)} analyses</span>
					</div>

					<div class="class-student-card-footer">
						<span class="text-sm text-muted">${student.last_predicted_at ? `Updated ${formatRelativeTime(student.last_predicted_at)}` : 'No updates yet'}</span>
						<span class="class-card-open">Open Profile <i data-lucide="arrow-right"></i></span>
					</div>
				</a>
			`;
		})
		.join('');
}

function renderScheduleDays(scheduleDays) {
	const days = Array.isArray(scheduleDays) ? scheduleDays : [];
	if (!days.length) {
		return '<span class="badge badge-gray">No schedule assigned</span>';
	}

	return days.map((day) => `<span class="badge badge-teal">${safeText(day)}</span>`).join('');
}

function renderErrorState(appElement, error) {
	const pageContent = `
		<div class="dashboard-container class-detail-page">
			<section class="card classes-empty-state">
				<i data-lucide="alert-triangle"></i>
				<h3>Unable to load class</h3>
				<p class="text-muted">${safeText(error.message || 'Please try again in a moment.')}</p>
				<a href="#/classes" class="btn btn-secondary">Back to Classes</a>
			</section>
		</div>
	`;

	appElement.innerHTML = createShell(pageContent, { topbarTitle: 'Class Detail' });
	attachShellHandlers();

	if (window.lucide && typeof window.lucide.createIcons === 'function') {
		window.lucide.createIcons();
	}
}

function getMoodBadgeClass(mood) {
	const normalized = String(mood || '').toLowerCase();
	if (normalized.includes('happy')) return 'badge-success';
	if (normalized.includes('sad')) return 'badge-warning';
	return 'badge-info';
}

function getInitials(name) {
	const parts = String(name || '').trim().split(/\s+/).filter(Boolean);
	if (!parts.length) return '?';
	return parts.slice(0, 2).map((part) => part.charAt(0).toUpperCase()).join('');
}

function formatDateShort(value) {
	if (!value) return 'N/A';
	const date = new Date(value);
	if (Number.isNaN(date.getTime())) return 'N/A';
	const day = String(date.getDate()).padStart(2, '0');
	const months = ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec'];
	return `${day}-${months[date.getMonth()]}-${String(date.getFullYear()).slice(-2)}`;
}

function formatRelativeTime(value) {
	if (!value) return 'N/A';
	const timestamp = new Date(value).getTime();
	if (Number.isNaN(timestamp)) return 'N/A';
	const delta = Date.now() - timestamp;
	const minute = 60 * 1000;
	const hour = 60 * minute;
	const day = 24 * hour;
	if (delta < minute) return 'Just now';
	if (delta < hour) return `${Math.floor(delta / minute)}m ago`;
	if (delta < day) return `${Math.floor(delta / hour)}h ago`;
	return `${Math.floor(delta / day)}d ago`;
}

function safeText(value) {
	const text = value == null ? '' : String(value);
	return text
		.replaceAll('&', '&amp;')
		.replaceAll('<', '&lt;')
		.replaceAll('>', '&gt;')
		.replaceAll('"', '&quot;')
		.replaceAll("'", '&#39;');
}